core/peerset/tests/fuzz.rs

```diff
@@ -0,0 +1,141 @@
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use futures::prelude::*;
+use libp2p::PeerId;
+use rand::distributions::{Distribution, Uniform, WeightedIndex};
+use rand::seq::IteratorRandom;
+use std::{collections::HashMap, collections::HashSet, iter};
+use substrate_peerset::{IncomingIndex, Message, PeersetConfig, Peerset};
+
+#[test]
+fn run() {
+	for _ in 0..25 {
+		test_once();
+	}
+}
+
+fn test_once() {
+	// PRNG to use.
+	let mut rng = rand::thread_rng();
+
+	// Nodes that the peerset knows about.
+	let mut known_nodes = HashSet::<PeerId>::new();
+	// Nodes that we have reserved. Always a subset of `known_nodes`.
+	let mut reserved_nodes = HashSet::<PeerId>::new();
+
+	let (mut peerset, peerset_handle) = Peerset::from_config(PeersetConfig {
+		bootnodes: (0 .. Uniform::new_inclusive(0, 4).sample(&mut rng)).map(|_| {
+			let id = PeerId::random();
+			known_nodes.insert(id.clone());
+			id
+		}).collect(),
+		reserved_nodes: (0 .. Uniform::new_inclusive(0, 2).sample(&mut rng)).map(|_| {
+			let id = PeerId::random();
+			known_nodes.insert(id.clone());
+			reserved_nodes.insert(id.clone());
+			id
+		}).collect(),
+		reserved_only: Uniform::new_inclusive(0, 10).sample(&mut rng) == 0,
+		in_peers: Uniform::new_inclusive(0, 25).sample(&mut rng),
+		out_peers: Uniform::new_inclusive(0, 25).sample(&mut rng),
+	});
+
+	tokio::runtime::current_thread::Runtime::new().unwrap().block_on(futures::future::poll_fn(move || -> Result<_, ()> {
+		// List of nodes the user of `peerset` assumes it's connected to. Always a subset of
+		// `known_nodes`.
+		let mut connected_nodes = HashSet::<PeerId>::new();
+		// List of nodes the user of `peerset` called `incoming` with and that haven't been
+		// accepted or rejected yet.
+		let mut incoming_nodes = HashMap::<IncomingIndex, PeerId>::new();
+		// Next id for incoming connections.
+		let mut next_incoming_id = IncomingIndex(0);
+
+		// Perform a certain number of actions while checking that the state is consistent. If we
+		// reach the end of the loop, the run has succeeded.
+		for _ in 0 .. 1000 {
+			// Each of these weights corresponds to an action that we may perform.
+			let action_weights = [150, 90, 90, 30, 30, 1, 1, 4, 4];
+			match WeightedIndex::new(&action_weights).unwrap().sample(&mut rng) {
+				// If we generate 0, poll the peerset.
+				0 => match peerset.poll().unwrap() {
+					Async::Ready(Some(Message::Connect(id))) => {
+						if let Some(id) = incoming_nodes.iter().find(|(_, v)| **v == id).map(|(&id, _)| id) {
+							incoming_nodes.remove(&id);
+						}
+						assert!(connected_nodes.insert(id));
```

Comment:
I think you might want to change this to `assert!(connected_nodes.insert(id) || known_nodes.remove(id));`, to account for an `Accept` message being received after `discovered` was called?

Reply:
`known_nodes` contains all the nodes that have been discovered in the past, not just nodes waiting for connection.